Can't find column in projection

This query:

```sql
SELECT
                    commit_hash,
                    n_repos,
                    s.repository_id AS main_repository_id,
                    repository_ids
                FROM (
                    SELECT
                        commit_hash,
                        COUNT(*) n_repos,
                        MAX(STRUCT(history_index, repository_id)) AS s,
                        COLLECT_SET(repository_id) AS repository_ids
                    FROM ref_commits
                    NATURAL JOIN commits
                    WHERE
                        history_index != 1
                        AND ref_name LIKE 'refs/heads/HEAD/%'
                        AND SIZE(parse_commit_parents(commit_parents)) == 0
                    GROUP BY commit_hash
                ) AS q
                WHERE n_repos > 1
                ORDER BY n_repos DESC
```

produces this error:

```
ObjectHashAggregate(keys=[commit_hash#9], functions=[count(1), max(named_struct(history_index, history_index#11L, repository_id, repository_id#8)), collect_set(repository_id#8, 0, 0)], output=[commit_hash#9, n_repos#92L, s#93, repository_ids#94])
+- Exchange hashpartitioning(commit_hash#9, 200)
   +- ObjectHashAggregate(keys=[commit_hash#9], functions=[partial_count(1), partial_max(named_struct(history_index, history_index#11L, repository_id, repository_id#8)), partial_collect_set(repository_id#8, 0, 0)], output=[commit_hash#9, count#120L, max#121, buf#122])
      +- *(1) !Project [repository_id#8, commit_hash#9, history_index#11L]
         +- *(1) Filter (size(UDF:parse_commit_parents(commit_parents#26)) = 0)
            +- *(1) DataSourceV2Scan [repository_id#8, commit_hash#9, commit_parents#26], DefaultReader(List(GitbaseServer(localhost:3308,root,)),StructType(StructField(repository_id,StringType,true), StructField(commit_hash,StringType,true), StructField(commit_parents,BinaryType,true)),Project(List(repository_id#8, commit_hash#9, commit_parents#26),Join(Table(ref_commits),Table(commits),Some(((repository_id#8 = repository_id#16) && (commit_hash#9 = commit_hash#17))))))

org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:
ObjectHashAggregate(keys=[commit_hash#9], functions=[count(1), max(named_struct(history_index, history_index#11L, repository_id, repository_id#8)), collect_set(repository_id#8, 0, 0)], output=[commit_hash#9, n_repos#92L, s#93, repository_ids#94])
+- Exchange hashpartitioning(commit_hash#9, 200)
   +- ObjectHashAggregate(keys=[commit_hash#9], functions=[partial_count(1), partial_max(named_struct(history_index, history_index#11L, repository_id, repository_id#8)), partial_collect_set(repository_id#8, 0, 0)], output=[commit_hash#9, count#120L, max#121, buf#122])
      +- *(1) !Project [repository_id#8, commit_hash#9, history_index#11L]
         +- *(1) Filter (size(UDF:parse_commit_parents(commit_parents#26)) = 0)
            +- *(1) DataSourceV2Scan [repository_id#8, commit_hash#9, commit_parents#26], DefaultReader(List(GitbaseServer(localhost:3308,root,)),StructType(StructField(repository_id,StringType,true), StructField(commit_hash,StringType,true), StructField(commit_parents,BinaryType,true)),Project(List(repository_id#8, commit_hash#9, commit_parents#26),Join(Table(ref_commits),Table(commits),Some(((repository_id#8 = repository_id#16) && (commit_hash#9 = commit_hash#17))))))

	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.execution.aggregate.ObjectHashAggregateExec.doExecute(ObjectHashAggregateExec.scala:100)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.InputAdapter.inputRDDs(WholeStageCodegenExec.scala:371)
	at org.apache.spark.sql.execution.FilterExec.inputRDDs(basicPhysicalOperators.scala:121)
	at org.apache.spark.sql.execution.ProjectExec.inputRDDs(basicPhysicalOperators.scala:41)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:605)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.TakeOrderedAndProjectExec.executeCollect(limit.scala:135)
	at org.apache.spark.sql.Dataset.org$apache$spark$sql$Dataset$$collectFromPlan(Dataset.scala:3273)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2484)
	at org.apache.spark.sql.Dataset$$anonfun$head$1.apply(Dataset.scala:2484)
	at org.apache.spark.sql.Dataset$$anonfun$52.apply(Dataset.scala:3254)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:77)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3253)
	at org.apache.spark.sql.Dataset.head(Dataset.scala:2484)
	at org.apache.spark.sql.Dataset.take(Dataset.scala:2698)
	at org.apache.spark.sql.Dataset.showString(Dataset.scala:254)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:725)
	at org.apache.spark.sql.Dataset.show(Dataset.scala:702)
	at tech.sourced.gitbase.spark.DefaultSourceSpec$$anonfun$12.apply$mcV$sp(DefaultSourceSpec.scala:343)
	at tech.sourced.gitbase.spark.DefaultSourceSpec$$anonfun$12.apply(DefaultSourceSpec.scala:343)
	at tech.sourced.gitbase.spark.DefaultSourceSpec$$anonfun$12.apply(DefaultSourceSpec.scala:343)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FlatSpecLike$$anon$1.apply(FlatSpecLike.scala:1682)
	at org.scalatest.TestSuite$class.withFixture(TestSuite.scala:196)
	at org.scalatest.FlatSpec.withFixture(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$class.invokeWithFixture$1(FlatSpecLike.scala:1679)
	at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
	at org.scalatest.FlatSpecLike$$anonfun$runTest$1.apply(FlatSpecLike.scala:1692)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:289)
	at org.scalatest.FlatSpecLike$class.runTest(FlatSpecLike.scala:1692)
	at org.scalatest.FlatSpec.runTest(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
	at org.scalatest.FlatSpecLike$$anonfun$runTests$1.apply(FlatSpecLike.scala:1750)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:396)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:373)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:410)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:384)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:384)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:379)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:461)
	at org.scalatest.FlatSpecLike$class.runTests(FlatSpecLike.scala:1750)
	at org.scalatest.FlatSpec.runTests(FlatSpec.scala:1685)
	at org.scalatest.Suite$class.run(Suite.scala:1147)
	at org.scalatest.FlatSpec.org$scalatest$FlatSpecLike$$super$run(FlatSpec.scala:1685)
	at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
	at org.scalatest.FlatSpecLike$$anonfun$run$1.apply(FlatSpecLike.scala:1795)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:521)
	at org.scalatest.FlatSpecLike$class.run(FlatSpecLike.scala:1795)
	at tech.sourced.gitbase.spark.DefaultSourceSpec.org$scalatest$BeforeAndAfterAll$$super$run(DefaultSourceSpec.scala:4)
	at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:213)
	at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:210)
	at tech.sourced.gitbase.spark.DefaultSourceSpec.run(DefaultSourceSpec.scala:4)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1346)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$1.apply(Runner.scala:1340)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1340)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1011)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1010)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1506)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1010)
	at org.scalatest.tools.Runner$.run(Runner.scala:850)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:131)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)
Caused by: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:
Exchange hashpartitioning(commit_hash#9, 200)
+- ObjectHashAggregate(keys=[commit_hash#9], functions=[partial_count(1), partial_max(named_struct(history_index, history_index#11L, repository_id, repository_id#8)), partial_collect_set(repository_id#8, 0, 0)], output=[commit_hash#9, count#120L, max#121, buf#122])
   +- *(1) !Project [repository_id#8, commit_hash#9, history_index#11L]
      +- *(1) Filter (size(UDF:parse_commit_parents(commit_parents#26)) = 0)
         +- *(1) DataSourceV2Scan [repository_id#8, commit_hash#9, commit_parents#26], DefaultReader(List(GitbaseServer(localhost:3308,root,)),StructType(StructField(repository_id,StringType,true), StructField(commit_hash,StringType,true), StructField(commit_parents,BinaryType,true)),Project(List(repository_id#8, commit_hash#9, commit_parents#26),Join(Table(ref_commits),Table(commits),Some(((repository_id#8 = repository_id#16) && (commit_hash#9 = commit_hash#17))))))

	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.doExecute(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.aggregate.ObjectHashAggregateExec$$anonfun$doExecute$1.apply(ObjectHashAggregateExec.scala:105)
	at org.apache.spark.sql.execution.aggregate.ObjectHashAggregateExec$$anonfun$doExecute$1.apply(ObjectHashAggregateExec.scala:100)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	... 83 more
Caused by: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: execute, tree:
ObjectHashAggregate(keys=[commit_hash#9], functions=[partial_count(1), partial_max(named_struct(history_index, history_index#11L, repository_id, repository_id#8)), partial_collect_set(repository_id#8, 0, 0)], output=[commit_hash#9, count#120L, max#121, buf#122])
+- *(1) !Project [repository_id#8, commit_hash#9, history_index#11L]
   +- *(1) Filter (size(UDF:parse_commit_parents(commit_parents#26)) = 0)
      +- *(1) DataSourceV2Scan [repository_id#8, commit_hash#9, commit_parents#26], DefaultReader(List(GitbaseServer(localhost:3308,root,)),StructType(StructField(repository_id,StringType,true), StructField(commit_hash,StringType,true), StructField(commit_parents,BinaryType,true)),Project(List(repository_id#8, commit_hash#9, commit_parents#26),Join(Table(ref_commits),Table(commits),Some(((repository_id#8 = repository_id#16) && (commit_hash#9 = commit_hash#17))))))

	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.execution.aggregate.ObjectHashAggregateExec.doExecute(ObjectHashAggregateExec.scala:100)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec.prepareShuffleDependency(ShuffleExchangeExec.scala:92)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:128)
	at org.apache.spark.sql.execution.exchange.ShuffleExchangeExec$$anonfun$doExecute$1.apply(ShuffleExchangeExec.scala:119)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	... 93 more
Caused by: org.apache.spark.sql.catalyst.errors.package$TreeNodeException: Binding attribute, tree: history_index#11L
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:56)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1.applyOrElse(BoundAttribute.scala:91)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1.applyOrElse(BoundAttribute.scala:90)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$2.apply(TreeNode.scala:267)
	at org.apache.spark.sql.catalyst.trees.TreeNode$$anonfun$2.apply(TreeNode.scala:267)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(TreeNode.scala:70)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transformDown(TreeNode.scala:266)
	at org.apache.spark.sql.catalyst.trees.TreeNode.transform(TreeNode.scala:256)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$.bindReference(BoundAttribute.scala:90)
	at org.apache.spark.sql.execution.ProjectExec$$anonfun$5.apply(basicPhysicalOperators.scala:59)
	at org.apache.spark.sql.execution.ProjectExec$$anonfun$5.apply(basicPhysicalOperators.scala:59)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.immutable.List.map(List.scala:296)
	at org.apache.spark.sql.execution.ProjectExec.doConsume(basicPhysicalOperators.scala:59)
	at org.apache.spark.sql.execution.CodegenSupport$class.consume(WholeStageCodegenExec.scala:181)
	at org.apache.spark.sql.execution.FilterExec.consume(basicPhysicalOperators.scala:85)
	at org.apache.spark.sql.execution.FilterExec.doConsume(basicPhysicalOperators.scala:206)
	at org.apache.spark.sql.execution.CodegenSupport$class.consume(WholeStageCodegenExec.scala:181)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanExec.consume(DataSourceV2ScanExec.scala:37)
	at org.apache.spark.sql.execution.ColumnarBatchScan$class.produceRows(ColumnarBatchScan.scala:166)
	at org.apache.spark.sql.execution.ColumnarBatchScan$class.doProduce(ColumnarBatchScan.scala:80)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanExec.doProduce(DataSourceV2ScanExec.scala:37)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:88)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.CodegenSupport$class.produce(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.datasources.v2.DataSourceV2ScanExec.produce(DataSourceV2ScanExec.scala:37)
	at org.apache.spark.sql.execution.FilterExec.doProduce(basicPhysicalOperators.scala:125)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:88)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.CodegenSupport$class.produce(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.FilterExec.produce(basicPhysicalOperators.scala:85)
	at org.apache.spark.sql.execution.ProjectExec.doProduce(basicPhysicalOperators.scala:45)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:88)
	at org.apache.spark.sql.execution.CodegenSupport$$anonfun$produce$1.apply(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.CodegenSupport$class.produce(WholeStageCodegenExec.scala:83)
	at org.apache.spark.sql.execution.ProjectExec.produce(basicPhysicalOperators.scala:35)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doCodeGen(WholeStageCodegenExec.scala:524)
	at org.apache.spark.sql.execution.WholeStageCodegenExec.doExecute(WholeStageCodegenExec.scala:576)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:131)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$execute$1.apply(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.SparkPlan$$anonfun$executeQuery$1.apply(SparkPlan.scala:155)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:152)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:127)
	at org.apache.spark.sql.execution.aggregate.ObjectHashAggregateExec$$anonfun$doExecute$1.apply(ObjectHashAggregateExec.scala:105)
	at org.apache.spark.sql.execution.aggregate.ObjectHashAggregateExec$$anonfun$doExecute$1.apply(ObjectHashAggregateExec.scala:100)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	... 104 more
Caused by: java.lang.RuntimeException: Couldn't find history_index#11L in [repository_id#8,commit_hash#9,commit_parents#26]
	at scala.sys.package$.error(package.scala:27)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1$$anonfun$applyOrElse$1.apply(BoundAttribute.scala:97)
	at org.apache.spark.sql.catalyst.expressions.BindReferences$$anonfun$bindReference$1$$anonfun$applyOrElse$1.apply(BoundAttribute.scala:91)
	at org.apache.spark.sql.catalyst.errors.package$.attachTree(package.scala:52)
	... 162 more
```

This is not related to #38, though. At first sight, I can imagine it's because we push down another project with the stuff that can be pushed down to gitbase and we keep the other udf without moving also the required columns there in [here](https://github.com/src-d/gitbase-spark-connector/pull/51/files#diff-8daa453aa8f593cfb55a13d1d8294bb1R140)

--- a/src/main/scala/tech/sourced/gitbase/spark/rule/PushdownJoins.scala
+++ b/src/main/scala/tech/sourced/gitbase/spark/rule/PushdownJoins.scala
@@ -170,27 +170,44 @@ private[rule] object JoinOptimizer extends Logging {
       case None =>
     }
 
-    // Check if the Join contains all valid Nodes
-    val jd: Seq[JoinData] = j.map {
-      case jm@logical.Join(_, _, _, condition) =>
-        if (jm == j) {
-          JoinData(conditions = condition, valid = true)
-        } else {
-          logUnableToOptimize(s"Invalid node: $jm")
-          JoinData()
-        }
+    val left = getLeafJoinData(j.left)
+    val right = getLeafJoinData(j.right)
+    mergeJoinData(Seq(
+      JoinData(valid = true, conditions = j.condition),
+      left,
+      right
+    ))
+  }
+
+  private def getLeafJoinData(p: LogicalPlan): JoinData = {
+    var filter: Option[Expression] = None
+    var valid = true
+    var project: Seq[NamedExpression] = Seq()
+    var source: Option[Node] = None
+    var attributes: Seq[AttributeReference] = Seq()
+    var servers: Seq[GitbaseServer] = Seq()
+
+    p.map {
+      case jm@logical.Join(_, _, _, _) =>
+        logUnableToOptimize(s"Invalid node: $jm")
+        valid = false
       case logical.Filter(cond, _) =>
-        JoinData(filter = Some(cond), valid = true)
+        filter = Some(cond)
       case logical.Project(namedExpressions, _) =>
-        JoinData(project = namedExpressions, valid = true)
-      case DataSourceV2Relation(out, DefaultReader(servers, _, source)) =>
-        JoinData(Some(source), attributes = out, servers = servers, valid = true)
+        project = namedExpressions
+      case DataSourceV2Relation(out, DefaultReader(srvs, _, src)) =>
+        source = Some(src)
+        if (project.isEmpty) {
+          project = out
+        }
+        attributes = out
+        servers = srvs
       case other =>
         logUnableToOptimize(s"Invalid node: $other")
-        JoinData()
+        valid = false
     }
 
-    mergeJoinData(jd)
+    JoinData(source, None, filter, project, attributes, servers, valid)
   }
 
   private def getRelationTables(left: DataSourceV2Relation,

--- a/src/main/scala/tech/sourced/gitbase/spark/rule/PushdownTree.scala
+++ b/src/main/scala/tech/sourced/gitbase/spark/rule/PushdownTree.scala
@@ -163,7 +163,7 @@ object PushdownTree extends Rule[LogicalPlan] {
 
           supported = supportedExp
           unsupported = Some(unsupportedNamedExp)
-        // this expression can be handled and it have no children
+        // this expression can be handled and it has no children
         case e if canBeHandled(e :: Nil) && unsupported.isEmpty =>
           val namedExp = toNamedExpression(e)
           supported = Seq(namedExp)

--- a/src/main/scala/tech/sourced/gitbase/spark/rule/PushdownTree.scala
+++ b/src/main/scala/tech/sourced/gitbase/spark/rule/PushdownTree.scala
@@ -13,7 +13,7 @@ import tech.sourced.gitbase.spark._
 
 object PushdownTree extends Rule[LogicalPlan] {
 
-  /** @inheritdoc*/
+  /** @inheritdoc */
   def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
     case logical.Project(exp1, logical.Project(exp2, child))
       if containsDuplicates(exp1 ++ exp2) =>
@@ -113,7 +113,7 @@ object PushdownTree extends Rule[LogicalPlan] {
   private def toNamedExpression(expression: Expression): NamedExpression = {
     expression match {
       case e: NamedExpression => e
-      case e => Alias(e, e.toString())()
+      case e => Alias(e, e.sql)()
     }
   }
 

--- a/src/test/scala/tech/sourced/gitbase/spark/DefaultSourceSpec.scala
+++ b/src/test/scala/tech/sourced/gitbase/spark/DefaultSourceSpec.scala
@@ -210,11 +210,11 @@ class DefaultSourceSpec extends BaseGitbaseSpec {
   }
 
   it should "get commit parents using parse_commit_parents" in {
-    val result = spark.sql("SELECT repository_id, parse_commit_parents(commit_parents) parents" +
+    val result = spark.sql("SELECT repository_id, parse_commit_parents(commit_parents)" +
       " FROM ref_commits" +
       " NATURAL JOIN commits" +
       " WHERE ref_name LIKE 'refs/heads/HEAD/%' AND history_index = 0" +
-      " ORDER BY repository_id, parents")
+      " ORDER BY repository_id, 2")
       .collect()
       .map(row => (row(0), row(1).asInstanceOf[Seq[String]]))
 

--- a/src/test/scala/tech/sourced/gitbase/spark/DefaultSourceSpec.scala
+++ b/src/test/scala/tech/sourced/gitbase/spark/DefaultSourceSpec.scala
@@ -249,4 +249,34 @@ class DefaultSourceSpec extends BaseGitbaseSpec {
       """.stripMargin)
     df.show(20, false)
   }
+
+  it should "approximately detect forks" in {
+    val result = spark.sql(
+      """SELECT
+                     commit_hash,
+                     n_repos,
+                     s.repository_id AS main_repository_id,
+                     repository_ids
+                 FROM (
+                     SELECT
+                         commit_hash,
+                         COUNT(*) n_repos,
+                         MAX(STRUCT(history_index, repository_id)) AS s,
+                         COLLECT_SET(repository_id) AS repository_ids
+                     FROM ref_commits
+                     NATURAL JOIN commits
+                     WHERE
+                         history_index != 1
+                         AND ref_name LIKE 'refs/heads/HEAD/%'
+                         AND SIZE(parse_commit_parents(commit_parents)) == 0
+                     GROUP BY commit_hash
+                 ) AS q
+                 WHERE n_repos > 1
+                 ORDER BY n_repos DESC""").collect().map(r => (r(0), r(1)))
+
+    result should equal(Array(
+      ("fff7062de8474d10a67d417ccea87ba6f58ca81d", 2),
+      ("fff840f8784ef162dc83a1465fc5763d890b68ba", 2)
+    ))
+  }
 }

--- a/src/test/scala/tech/sourced/gitbase/spark/rule/PushdownTreeSpec.scala
+++ b/src/test/scala/tech/sourced/gitbase/spark/rule/PushdownTreeSpec.scala
@@ -1,7 +1,9 @@
 package tech.sourced.gitbase.spark.rule
 
 import org.apache.spark.sql.catalyst.expressions.{
-  Alias, Ascending, Attribute, CaseKeyWhen, Descending, EqualTo, Literal, SortOrder}
+  Alias, Ascending, Attribute, CaseKeyWhen,
+  Descending, EqualTo, Literal, SortOrder
+}
 import org.apache.spark.sql.catalyst.plans.logical
 import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Relation
 import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
@@ -192,4 +194,5 @@ class PushdownTreeSpec extends BaseRuleSpec {
 
     PushdownTree(node).isInstanceOf[logical.GlobalLimit] should be(true)
   }
+
 }

